src/GitHub.App/Models/RepositoryHosts.cs

```diff
@@ -45,78 +40,20 @@ public class RepositoryHosts : ReactiveObject, IRepositoryHosts
             GitHubHost = DisconnectedRepositoryHost;
             EnterpriseHost = DisconnectedRepositoryHost;
 
-            var persistEntepriseHostObs = this.WhenAny(x => x.EnterpriseHost, x => x.Value)
-                .Skip(1)  // The first value will be null or something already in the db
-                .SelectMany(enterpriseHost =>
-                {
-                    if (!enterpriseHost.IsLoggedIn)
-                    {
-                        return sharedCache.UserAccount
-                            .InvalidateObject<Uri>(EnterpriseHostApiBaseUriCacheKey)
-                            .Catch<Unit, Exception>(ex =>
-                            {
-                                log.Warn("Failed to invalidate enterprise host uri", ex);
-                                return Observable.Return(Unit.Default);
-                            });
-                    }
-
-                    return sharedCache.UserAccount
-                        .InsertObject(EnterpriseHostApiBaseUriCacheKey, enterpriseHost.Address.ApiUri)
-                        .Catch<Unit, Exception>(ex =>
-                        {
-                            log.Warn("Failed to persist enterprise host uri", ex);
-                            return Observable.Return(Unit.Default);
-                        });
-                });
-
             isLoggedInToAnyHost = this.WhenAny(
                 x => x.GitHubHost.IsLoggedIn,
                 x => x.EnterpriseHost.IsLoggedIn,
                 (githubLoggedIn, enterpriseLoggedIn) => githubLoggedIn.Value || enterpriseLoggedIn.Value)
                 .ToProperty(this, x => x.IsLoggedInToAnyHost);
 
-            // This part is strictly to support having the IConnectionManager request that a connection
-            // be logged in. It doesn't know about hosts or load anything reactive, so it gets
-            // informed of logins by an observable returned by the event
-            connectionManager.DoLogin += RunLoginHandler;
-
-            // monitor the list of connections so we can log out hosts when connections are removed
-            disposables.Add(
-                connectionManager.Connections.CreateDerivedCollection(x => x)
-                .ItemsRemoved
-                .SelectMany(async x =>
-                {
-                    var host = LookupHost(x.HostAddress);
-                    if (host.Address != x.HostAddress)
-                        host = await RepositoryHostFactory.Create(x.HostAddress);
-                    return host;
-                })
-                .Select(h => LogOut(h))
-                .Merge().ToList().Select(_ => Unit.Default).Subscribe());
-
+            connectionManager.ConnectionCreated = ConnectionAdded;
```

Comment:
This is a temporary hack, that performs the same purpose as `connectionManager.DoLogin`, but renamed (as it no longer does a login) and using `Task` instead of an observable.

It's needed because creating a RepositoryHost is async meaning that for a short time a connection can exist without
a repository host, but other older parts of the code get confused by this. This callback allows RepositoryHosts to hook into the creation of the connection to make sure a host is available by the time the connection is made available.

The longer-term goal is to remove `RepositoryHost(s)` so this can go away.